mantidproject/mantid
# Mantid calibration for Mandi as good as that from ISAW

Description of work.
Two problems were found in SCD calibrations in Mantid.  First the B matrix was not fixed and second there were some errors in the analytic derivatives.  It now uses numerical derivatives and only the U matrix is allowed to change.

**To test:**



Get mandi input files from dropbox and use python script to check that indexing error from Mantid calibration is as good as one from ISAW.
https://www.dropbox.com/s/k6f6auy4ghx0p4q/natrolite_Niggli_0.12.peaks?dl=0
https://www.dropbox.com/s/3r0n6u5qa16gqbb/natrolite_Niggli.mat?dl=0
https://www.dropbox.com/s/tg8vklot51cap85/natrolite_Niggli_0.12_2016.peaks?dl=0

``` python
import time

LoadIsawPeaks(Filename='natrolite_Niggli_0.12.peaks', OutputWorkspace='natrolite')
LoadIsawUB(InputWorkspace='natrolite', Filename='natrolite_Niggli.mat')
CloneWorkspace(InputWorkspace="natrolite",OutputWorkspace="natrolite_orig")
CloneWorkspace(InputWorkspace="natrolite_orig",OutputWorkspace="natrolite_Simplex")
num,err=IndexPeaks(PeaksWorkspace='natrolite_orig',RoundHKLs=False)
print "Original peaks file: number indexed, error indexing = ", num,err
wksp=mtd['natrolite_orig']
lattice = wksp.sample().getOrientedLattice()
print "lattice, errors = ", lattice.a(),lattice.b(),lattice.c(),lattice.alpha(),lattice.beta(),lattice.gamma(),lattice.errora(),lattice.errorb(),lattice.errorc(),lattice.erroralpha(),lattice.errorbeta(),lattice.errorgamma()
print "U = ",lattice.getU()
print "B = ", lattice.getB()

#Levenberg-Marquardt optimization
start_time = time.time()
SCDCalibratePanels(PeakWorkspace='natrolite',
    useL0=True, usetimeOffset=False,usePanelWidth=True, usePanelHeight=True, usePanelPosition=True, usePanelOrientation=False, RotateCenters=True,
    AllowSampleShift=False, DetCalFilename='mandiLM.DetCal', ResultWorkspace='LMResults',
    ColWorkspace='LMCol', RowWorkspace='LMRow', TofWorkspace='LMTof', NumIterations=20)
print "minutes:  ", str((time.time()-start_time)/60)
num,err=IndexPeaks(PeaksWorkspace='natrolite',RoundHKLs=False)
print "MANTID optimized peaks file: number indexed, error indexing = ", num,err
wksp=mtd['natrolite']
lattice = wksp.sample().getOrientedLattice()
print "lattice, errors = ", lattice.a(),lattice.b(),lattice.c(),lattice.alpha(),lattice.beta(),lattice.gamma(),lattice.errora(),lattice.errorb(),lattice.errorc(),lattice.erroralpha(),lattice.errorbeta(),lattice.errorgamma()
print "U = ",lattice.getU()
print "B = ", lattice.getB()

#Simplex optimization
start_time = time.time()
SCDCalibratePanels(PeakWorkspace='natrolite_Simplex',
    useL0=True, usetimeOffset=False,usePanelWidth=True, usePanelHeight=True, usePanelPosition=True, usePanelOrientation=False, RotateCenters=True,
    AllowSampleShift=False, DetCalFilename='mandiSimplex.DetCal', ResultWorkspace='SimplexResults',
    ColWorkspace='SimplexCol', RowWorkspace='SimplexRow', TofWorkspace='SimplexTof', NumIterations=100,Minimizer='Simplex')
print "minutes:  ", str((time.time()-start_time)/60)
num,err=IndexPeaks(PeaksWorkspace='natrolite_Simplex',RoundHKLs=False)
print "MANTID Simplex optimized peaks file: number indexed, error indexing = ", num,err
wksp=mtd['natrolite_Simplex']
lattice = wksp.sample().getOrientedLattice()
print "lattice, errors = ", lattice.a(),lattice.b(),lattice.c(),lattice.alpha(),lattice.beta(),lattice.gamma(),lattice.errora(),lattice.errorb(),lattice.errorc(),lattice.erroralpha(),lattice.errorbeta(),lattice.errorgamma()
print "U = ",lattice.getU()
print "B = ", lattice.getB()

#Isaw Levenberg-Marquardt optimization
LoadIsawPeaks(Filename='natrolite_Niggli_0.12_2016.peaks', OutputWorkspace='natrolite_ISAW')
LoadIsawUB(InputWorkspace='natrolite_ISAW', Filename='natrolite_Niggli.mat')
num,err=IndexPeaks(PeaksWorkspace='natrolite_ISAW',RoundHKLs=False)
print "ISAW optimized peaks file: number indexed, error indexing = ", num,err
wksp=mtd['natrolite_ISAW']
lattice = wksp.sample().getOrientedLattice()
print "lattice, errors = ", lattice.a(),lattice.b(),lattice.c(),lattice.alpha(),lattice.beta(),lattice.gamma(),lattice.errora(),lattice.errorb(),lattice.errorc(),lattice.erroralpha(),lattice.errorbeta(),lattice.errorgamma()
print "U = ",lattice.getU()
print "B = ", lattice.getB()
print lattice.getB()
```

Fixes #15644 



---
#### Reviewer

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):
##### Code Review
##### Functional Tests
- Does everything look good? Comment with the ship it emoji but don't merge. A member of `@mantidproject/gatekeepers` will take care of it.
